File: backend/app/routers/skills.py
"""Router Skills — personas YAML avec system prompt, outils et paramètres."""
from __future__ import annotations

import json
import logging
import re
from pathlib import Path

import aiofiles
import yaml
from fastapi import APIRouter, HTTPException, UploadFile, File
from pydantic import BaseModel

from app.config import settings

logger = logging.getLogger(__name__)
router = APIRouter()

# ------------------------------------------------------------------
# Schéma d'une skill
# ------------------------------------------------------------------

DEFAULT_SKILLS = [
    {
        "id": "default",
        "name": "Assistant général",
        "description": "Assistant polyvalent avec accès à tous les outils",
        "icon": "🤖",
        "system_prompt": (
            "Tu es OllamaStudio, un assistant développeur expert. "
            "Tu as accès à des outils pour lire/écrire des fichiers, exécuter des commandes shell, "
            "faire des recherches dans le code et les documents importés. "
            "Réponds toujours en français sauf si l'utilisateur utilise une autre langue. "
            "Sois précis, concis et propose du code de qualité production."
        ),
        "enabled_tools": None,  # None = tous les outils
        "temperature": 0.7,
        "max_tokens": 4096,
        "color": "#6366f1",
    },
    {
        "id": "code-review",
        "name": "Code Review",
        "description": "Expert en revue de code, sécurité et bonnes pratiques",
        "icon": "🔍",
        "system_prompt": (
            "Tu es un expert en revue de code senior. Analyse le code fourni selon ces critères : "
            "1) Sécurité (OWASP, injections, secrets exposés) "
            "2) Performance et complexité algorithmique "
            "3) Maintenabilité et respect des principes SOLID "
            "4) Couverture de tests "
            "Fournis des commentaires structurés avec exemples de correction."
        ),
        "enabled_tools": ["read_file", "grep_files", "list_files"],
        "temperature": 0.3,
        "max_tokens": 8192,
        "color": "#f59e0b",
    },
    {
        "id": "devops",
        "name": "DevOps & Infrastructure",
        "description": "Expert Docker, CI/CD, Linux, configuration serveur",
        "icon": "🏗️",
        "system_prompt": (
            "Tu es un ingénieur DevOps senior spécialisé sur Linux (Debian/Fedora), Docker, "
            "et les pipelines CI/CD. Tu aides à configurer des environnements, rédiger des Dockerfiles "
            "optimisés et des scripts d'automatisation. Respecte les bonnes pratiques de sécurité "
            "système (ANSSI, CIS Benchmarks) quand applicable."
        ),
        "enabled_tools": ["read_file", "write_file", "run_command", "list_files", "grep_files", "git_status"],
        "temperature": 0.5,
        "max_tokens": 4096,
        "color": "#10b981",
    },
    {
        "id": "rust-expert",
        "name": "Expert Rust",
        "description": "Développeur Rust avancé — ownership, lifetimes, async, performances",
        "icon": "🦀",
        "system_prompt": (
            "Tu es un expert Rust de niveau avancé. Tu maîtrises le borrow checker, les lifetimes, "
            "les traits, async/await avec Tokio, les FFI et les optimisations de performances. "
            "Tu proposes du code idiomatique Rust en suivant les conventions de la communauté. "
            "Tu expliques les concepts complexes avec des exemples concrets."
        ),
        "enabled_tools": ["read_file", "write_file", "run_command", "grep_files", "git_status"],
        "temperature": 0.4,
        "max_tokens": 8192,
        "color": "#ef4444",
    },
    {
        "id": "security-audit",
        "name": "Audit Sécurité",
        "description": "Analyse de sécurité, conformité ANSSI/NIS2, hardening",
        "icon": "🛡️",
        "system_prompt": (
            "Tu es un auditeur sécurité expert (ANSSI, NIS2, ISO 27001). "
            "Tu analyses les configurations, les Dockerfiles, les scripts et le code source "
            "pour identifier les vulnérabilités. Tu fournis des recommandations de hardening "
            "conformes aux référentiels ANSSI et aux benchmarks CIS. "
            "Structure tes rapports : Risque critique / Élevé / Moyen / Faible."
        ),
        "enabled_tools": ["read_file", "grep_files", "list_files", "search_documents"],
        "temperature": 0.2,
        "max_tokens": 8192,
        "color": "#8b5cf6",
    },
]


# ------------------------------------------------------------------
# Helpers I/O
# ------------------------------------------------------------------

def _skill_path(skill_id: str) -> Path:
    return settings.skills_dir / f"{skill_id}.yaml"


async def load_skill(skill_id: str) -> dict | None:
    """Charge une skill depuis le fichier YAML ou les defaults."""
    # Cherche dans les fichiers custom d'abord
    path = _skill_path(skill_id)
    if path.exists():
        async with aiofiles.open(path, encoding="utf-8") as f:
            content = await f.read()
        data = yaml.safe_load(content)
        data["id"] = skill_id
        return data

    # Cherche dans les defaults embarqués
    for s in DEFAULT_SKILLS:
        if s["id"] == skill_id:
            return dict(s)

    return None


async def seed_default_skills() -> None:
    """Écrit les skills par défaut dans le dossier data/skills/ si inexistants."""
    for skill in DEFAULT_SKILLS:
        path = _skill_path(skill["id"])
        if not path.exists():
            skill_data = {k: v for k, v in skill.items() if k != "id"}
            async with aiofiles.open(path, "w", encoding="utf-8") as f:
                await f.write(yaml.dump(skill_data, allow_unicode=True, default_flow_style=False))
    logger.info("Skills initialisées dans %s", settings.skills_dir)


# ------------------------------------------------------------------
# Endpoints
# ------------------------------------------------------------------

class SkillCreate(BaseModel):
    id: str
    name: str
    description: str = ""
    icon: str = "🤖"
    system_prompt: str
    enabled_tools: list[str] | None = None
    temperature: float = 0.7
    max_tokens: int = 4096
    color: str = "#6366f1"


@router.get("/")
async def list_skills() -> list[dict]:
    """Retourne toutes les skills (defaults + custom)."""
    skills: dict[str, dict] = {}

    # Defaults
    for s in DEFAULT_SKILLS:
        skills[s["id"]] = dict(s)

    # Custom (écrasent les defaults si même ID)
    for path in settings.skills_dir.glob("*.yaml"):
        skill_id = path.stem
        async with aiofiles.open(path, encoding="utf-8") as f:
            content = await f.read()
        data = yaml.safe_load(content) or {}
        data["id"] = skill_id
        skills[skill_id] = data

    return list(skills.values())


@router.get("/{skill_id}")
async def get_skill(skill_id: str) -> dict:
    skill = await load_skill(skill_id)
    if skill is None:
        raise HTTPException(status_code=404, detail="Skill introuvable")
    return skill


@router.post("/", status_code=201)
async def create_skill(body: SkillCreate) -> dict:
    path = _skill_path(body.id)
    if path.exists():
        raise HTTPException(status_code=409, detail="Skill existante — utilisez PUT pour modifier")
    data = body.model_dump(exclude={"id"})
    async with aiofiles.open(path, "w", encoding="utf-8") as f:
        await f.write(yaml.dump(data, allow_unicode=True, default_flow_style=False))
    return {"id": body.id, **data}


@router.put("/{skill_id}")
async def update_skill(skill_id: str, body: SkillCreate) -> dict:
    path = _skill_path(skill_id)
    data = body.model_dump(exclude={"id"})
    async with aiofiles.open(path, "w", encoding="utf-8") as f:
        await f.write(yaml.dump(data, allow_unicode=True, default_flow_style=False))
    return {"id": skill_id, **data}


@router.delete("/{skill_id}", status_code=204)
async def delete_skill(skill_id: str) -> None:
    # Interdit la suppression des skills par défaut
    if any(s["id"] == skill_id for s in DEFAULT_SKILLS):
        raise HTTPException(status_code=403, detail="Impossible de supprimer une skill par défaut")
    path = _skill_path(skill_id)
    if not path.exists():
        raise HTTPException(status_code=404, detail="Skill introuvable")
    path.unlink()


# ------------------------------------------------------------------
# Import JSON en masse (compatible Claude / format libre)
# ------------------------------------------------------------------

def _slugify(text: str) -> str:
    """Génère un ID slug à partir d'un nom."""
    slug = re.sub(r"[^\w\s-]", "", text.lower().strip())
    return re.sub(r"[\s_]+", "-", slug)[:64] or "imported-skill"


def _normalize_skill(raw: dict, index: int) -> dict | None:
    """Normalise une entrée JSON en skill OllamaStudio.

    Accepte les formats :
    - OllamaStudio natif (id, name, system_prompt, ...)
    - Claude (name, content/instructions/system_prompt, ...)
    - Format simplifié (name + prompt)
    """
    name = raw.get("name") or raw.get("title") or f"Skill importée #{index + 1}"
    skill_id = raw.get("id") or raw.get("slug") or _slugify(name)

    # Détecte le champ prompt principal
    system_prompt = (
        raw.get("system_prompt")
        or raw.get("content")
        or raw.get("instructions")
        or raw.get("prompt")
        or raw.get("system")
        or ""
    )
    if not system_prompt:
        return None

    # Outils activés
    enabled_tools = raw.get("enabled_tools")
    if enabled_tools is not None and not isinstance(enabled_tools, list):
        enabled_tools = None

    return {
        "id": skill_id,
        "name": name,
        "description": raw.get("description", ""),
        "icon": raw.get("icon") or raw.get("emoji") or "📦",
        "system_prompt": system_prompt,
        "enabled_tools": enabled_tools,
        "temperature": float(raw.get("temperature", 0.7)),
        "max_tokens": int(raw.get("max_tokens", 4096)),
        "color": raw.get("color", "#6366f1"),
    }


@router.post("/import", status_code=201)
async def import_skills(file: UploadFile = File(...)) -> dict:
    """Importe des skills depuis un fichier JSON.

    Accepte un tableau JSON de skills ou un objet unique.
    Compatible avec les exports Claude et les formats personnalisés.
    """
    if not file.filename or not file.filename.lower().endswith(".json"):
        raise HTTPException(status_code=400, detail="Fichier JSON attendu (.json)")

    content = await file.read()
    if len(content) > 5 * 1024 * 1024:
        raise HTTPException(status_code=413, detail="Fichier trop volumineux (max 5 Mo)")

    try:
        data = json.loads(content.decode("utf-8"))
    except (json.JSONDecodeError, UnicodeDecodeError) as e:
        raise HTTPException(status_code=400, detail=f"JSON invalide: {e}")

    # Accepte un tableau ou un objet unique
    if isinstance(data, dict):
        # Si c'est un objet avec une clé qui contient un tableau (ex: {"skills": [...]})
        for key in ("skills", "personas", "agents", "items", "data"):
            if key in data and isinstance(data[key], list):
                data = data[key]
                break
        else:
            data = [data]

    if not isinstance(data, list):
        raise HTTPException(status_code=400, detail="Format attendu: tableau JSON ou objet avec clé 'skills'")

    imported = []
    skipped = []
    errors = []

    for i, raw in enumerate(data):
        if not isinstance(raw, dict):
            errors.append({"index": i, "error": "Entrée non-objet ignorée"})
            continue

        skill = _normalize_skill(raw, i)
        if skill is None:
            skipped.append({"index": i, "name": raw.get("name", "?"), "reason": "Pas de prompt/contenu"})
            continue

        skill_id = skill["id"]
        path = _skill_path(skill_id)

        # Si l'ID existe déjà, suffit avec un compteur
        if path.exists():
            counter = 2
            while _skill_path(f"{skill_id}-{counter}").exists():
                counter += 1
            skill_id = f"{skill_id}-{counter}"
            skill["id"] = skill_id
            path = _skill_path(skill_id)

        try:
            yaml_data = {k: v for k, v in skill.items() if k != "id"}
            async with aiofiles.open(path, "w", encoding="utf-8") as f:
                await f.write(yaml.dump(yaml_data, allow_unicode=True, default_flow_style=False))
            imported.append({"id": skill_id, "name": skill["name"]})
        except Exception as e:
            errors.append({"index": i, "name": skill["name"], "error": str(e)})

    return {
        "imported": len(imported),
        "skipped": len(skipped),
        "errors": len(errors),
        "details": {
            "imported": imported,
            "skipped": skipped,
            "errors": errors,
        },
    }

File: backend-rs/src/routes/skills.rs
//! Router Skills — YAML-backed personas with system prompt, tools, and parameters.

use axum::extract::{Multipart, Path, State};
use axum::http::StatusCode;
use axum::Json;
use serde_json::{json, Value};

use crate::config::Config;
use crate::error::AppError;
use crate::models::skill::{default_skills, slugify, Skill, SkillCreate};
use crate::state::AppState;

// ── Helpers I/O ──

fn skill_path(config: &Config, skill_id: &str) -> std::path::PathBuf {
    config.skills_dir().join(format!("{skill_id}.yaml"))
}

/// Load a skill by ID — checks YAML file first, then defaults.
/// Exported for use by the chat router.
pub async fn load_skill(skill_id: &str, config: &Config) -> Option<Skill> {
    let path = skill_path(config, skill_id);
    if path.exists() {
        if let Ok(content) = tokio::fs::read_to_string(&path).await {
            if let Ok(mut skill) = serde_yaml::from_str::<Skill>(&content) {
                skill.id = skill_id.to_string();
                return Some(skill);
            }
        }
    }

    // Fall back to built-in defaults
    default_skills()
        .into_iter()
        .find(|s| s.id == skill_id)
}

/// Seed default skills to YAML files on startup.
pub async fn seed_default_skills(config: &Config) -> Result<(), std::io::Error> {
    tokio::fs::create_dir_all(config.skills_dir()).await?;

    for skill in default_skills() {
        let path = skill_path(config, &skill.id);
        if !path.exists() {
            let data = skill_to_yaml_value(&skill);
            let yaml_str = serde_yaml::to_string(&data)
                .unwrap_or_default();
            tokio::fs::write(&path, yaml_str.as_bytes()).await?;
        }
    }
    tracing::info!("Skills initialis\u{00e9}es dans {}", config.skills_dir().display());
    Ok(())
}

// ── Endpoints ──

/// GET /api/skills/ — list all skills (defaults + custom YAML files).
pub async fn list_skills(
    State(state): State<AppState>,
) -> Result<Json<Vec<Value>>, AppError> {
    let mut skills: std::collections::HashMap<String, Skill> = std::collections::HashMap::new();

    // Load defaults first
    for s in default_skills() {
        skills.insert(s.id.clone(), s);
    }

    // Override / add from YAML files
    let skills_dir = state.config.skills_dir();
    if let Ok(mut entries) = tokio::fs::read_dir(&skills_dir).await {
        while let Ok(Some(entry)) = entries.next_entry().await {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) == Some("yaml") {
                let skill_id = path
                    .file_stem()
                    .and_then(|s| s.to_str())
                    .unwrap_or("")
                    .to_string();
                if let Ok(content) = tokio::fs::read_to_string(&path).await {
                    if let Ok(mut skill) = serde_yaml::from_str::<Skill>(&content) {
                        skill.id = skill_id.clone();
                        skills.insert(skill_id, skill);
                    }
                }
            }
        }
    }

    let result: Vec<Value> = skills
        .into_values()
        .map(|s| serde_json::to_value(s).unwrap_or_default())
        .collect();

    Ok(Json(result))
}

/// GET /api/skills/{id} — get a single skill.
pub async fn get_skill(
    State(state): State<AppState>,
    Path(skill_id): Path<String>,
) -> Result<Json<Value>, AppError> {
    let skill = load_skill(&skill_id, &state.config)
        .await
        .ok_or_else(|| AppError::NotFound("Skill introuvable".to_string()))?;
    Ok(Json(serde_json::to_value(skill).unwrap_or_default()))
}

/// POST /api/skills/ — create a new skill (write YAML).
pub async fn create_skill(
    State(state): State<AppState>,
    Json(body): Json<SkillCreate>,
) -> Result<(StatusCode, Json<Value>), AppError> {
    let path = skill_path(&state.config, &body.id);
    if path.exists() {
        return Err(AppError::Conflict(
            "Skill existante \u{2014} utilisez PUT pour modifier".to_string(),
        ));
    }

    let skill = skill_from_create(body.id.clone(), body);

    let data = skill_to_yaml_value(&skill);
    let yaml_str =
        serde_yaml::to_string(&data).map_err(|e| AppError::Internal(format!("YAML: {e}")))?;
    tokio::fs::write(&path, yaml_str.as_bytes()).await?;

    Ok((
        StatusCode::CREATED,
        Json(serde_json::to_value(&skill).unwrap_or_default()),
    ))
}

/// PUT /api/skills/{id} — update an existing skill (write YAML).
pub async fn update_skill(
    State(state): State<AppState>,
    Path(skill_id): Path<String>,
    Json(body): Json<SkillCreate>,
) -> Result<Json<Value>, AppError> {
    let path = skill_path(&state.config, &skill_id);

    let skill = skill_from_create(skill_id, body);

    let data = skill_to_yaml_value(&skill);
    let yaml_str =
        serde_yaml::to_string(&data).map_err(|e| AppError::Internal(format!("YAML: {e}")))?;
    tokio::fs::write(&path, yaml_str.as_bytes()).await?;

    Ok(Json(serde_json::to_value(&skill).unwrap_or_default()))
}

/// DELETE /api/skills/{id} — delete a custom skill (403 for defaults).
pub async fn delete_skill(
    State(state): State<AppState>,
    Path(skill_id): Path<String>,
) -> Result<StatusCode, AppError> {
    // Prevent deleting built-in skills
    if default_skills().iter().any(|s| s.id == skill_id) {
        return Err(AppError::Forbidden(
            "Impossible de supprimer une skill par d\u{00e9}faut".to_string(),
        ));
    }

    let path = skill_path(&state.config, &skill_id);
    if !path.exists() {
        return Err(AppError::NotFound("Skill introuvable".to_string()));
    }

    tokio::fs::remove_file(&path).await?;
    Ok(StatusCode::NO_CONTENT)
}

/// POST /api/skills/import — import skills from a JSON file (multipart).
/// Supports Claude, OllamaStudio, and generic formats.
pub async fn import_skills(
    State(state): State<AppState>,
    mut multipart: Multipart,
) -> Result<(StatusCode, Json<Value>), AppError> {
    state.debug.log("INFO", "skill", "Import skills depuis fichier");

    // Read the single file field
    let field = multipart
        .next_field()
        .await
        .map_err(|e| AppError::BadRequest(format!("Erreur multipart: {e}")))?
        .ok_or_else(|| AppError::BadRequest("Aucun fichier fourni".to_string()))?;

    let file_name = field
        .file_name()
        .unwrap_or("unknown")
        .to_string();

    if !file_name.to_lowercase().ends_with(".json") {
        return Err(AppError::BadRequest(
            "Fichier JSON attendu (.json)".to_string(),
        ));
    }

    let content_bytes = field
        .bytes()
        .await
        .map_err(|e| AppError::Internal(format!("Lecture fichier: {e}")))?;

    if content_bytes.len() > 5 * 1024 * 1024 {
        return Err(AppError::PayloadTooLarge(
            "Fichier trop volumineux (max 5 Mo)".to_string(),
        ));
    }

    let text = String::from_utf8_lossy(&content_bytes);
    let mut data: Value =
        serde_json::from_str(&text).map_err(|e| AppError::BadRequest(format!("JSON invalide: {e}")))?;

    // Accept array or object with known keys
    if let Some(obj) = data.as_object() {
        let mut found = false;
        for key in &["skills", "personas", "agents", "items", "data"] {
            if let Some(val) = obj.get(*key) {
                if val.is_array() {
                    // Direct array: {"skills": [...]}
                    data = val.clone();
                    found = true;
                    break;
                } else if let Some(inner_obj) = val.as_object() {
                    // Nested object: {"skills": {"public": [...], "user": [...], ...}}
                    // Flatten all sub-arrays into one list
                    let mut all_items = Vec::new();
                    for (_sub_key, sub_val) in inner_obj {
                        if let Some(arr) = sub_val.as_array() {
                            all_items.extend(arr.iter().cloned());
                        }
                    }
                    if !all_items.is_empty() {
                        data = Value::Array(all_items);
                        found = true;
                        break;
                    }
                }
            }
        }
        if !found {
            // Single object → wrap in array
            data = Value::Array(vec![data]);
        }
    }

    let items = data
        .as_array()
        .ok_or_else(|| {
            AppError::BadRequest(
                "Format attendu: tableau JSON ou objet avec cl\u{00e9} 'skills'".to_string(),
            )
        })?;

    let mut imported: Vec<Value> = Vec::new();
    let mut skipped: Vec<Value> = Vec::new();
    let mut errors: Vec<Value> = Vec::new();

    for (i, raw) in items.iter().enumerate() {
        let Some(obj) = raw.as_object() else {
            errors.push(json!({"index": i, "error": "Entr\u{00e9}e non-objet ignor\u{00e9}e"}));
            continue;
        };

        let Some(skill) = normalize_skill(obj, i) else {
            skipped.push(json!({
                "index": i,
                "name": obj.get("name").and_then(|v| v.as_str()).unwrap_or("?"),
                "reason": "Pas de prompt/contenu",
            }));
            continue;
        };

        let mut skill_id = skill.id.clone();
        let mut path = skill_path(&state.config, &skill_id);

        // Handle ID collision
        if path.exists() {
            let mut counter = 2u32;
            loop {
                let candidate = format!("{skill_id}-{counter}");
                let candidate_path = skill_path(&state.config, &candidate);
                if !candidate_path.exists() {
                    skill_id = candidate;
                    path = candidate_path;
                    break;
                }
                counter += 1;
            }
        }

        let mut final_skill = skill;
        final_skill.id = skill_id.clone();

        let data = skill_to_yaml_value(&final_skill);
        match serde_yaml::to_string(&data) {
            Ok(yaml_str) => match tokio::fs::write(&path, yaml_str.as_bytes()).await {
                Ok(()) => {
                    state.debug.log("INFO", "skill", &format!("Skill import\u{00e9}e: {skill_id} ({})", final_skill.name));
                    imported.push(json!({"id": skill_id, "name": final_skill.name}));
                }
                Err(e) => {
                    errors.push(json!({"index": i, "name": final_skill.name, "error": e.to_string()}));
                }
            },
            Err(e) => {
                errors.push(json!({"index": i, "name": final_skill.name, "error": e.to_string()}));
            }
        }
    }

    state.debug.log(
        "INFO",
        "skill",
        &format!(
            "Import termin\u{00e9}: {} import\u{00e9}s, {} ignor\u{00e9}s, {} erreurs",
            imported.len(),
            skipped.len(),
            errors.len()
        ),
    );

    Ok((
        StatusCode::CREATED,
        Json(json!({
            "imported": imported.len(),
            "skipped": skipped.len(),
            "errors": errors.len(),
            "details": {
                "imported": imported,
                "skipped": skipped,
                "errors": errors,
            },
        })),
    ))
}

// ── Internal helpers ──

/// Normalize a raw JSON object into a Skill, supporting Claude / OllamaStudio / generic formats.
fn normalize_skill(
    raw: &serde_json::Map<String, Value>,
    index: usize,
) -> Option<Skill> {
    let name = raw
        .get("name")
        .or(raw.get("title"))
        .and_then(|v| v.as_str())
        .unwrap_or(&format!("Skill import\u{00e9}e #{}", index + 1))
        .to_string();

    let skill_id = raw
        .get("id")
        .or(raw.get("slug"))
        .and_then(|v| v.as_str())
        .map(String::from)
        .unwrap_or_else(|| slugify(&name));

    // Detect the main prompt field — description accepted as last-resort fallback
    let system_prompt = [
        "system_prompt",
        "content",
        "instructions",
        "prompt",
        "system",
        "description",
    ]
    .iter()
    .find_map(|key| raw.get(*key).and_then(|v| v.as_str()).filter(|s| !s.is_empty()))
    .map(String::from)?;

    // enabled_tools
    let enabled_tools = raw
        .get("enabled_tools")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect::<Vec<String>>()
        });

    // If system_prompt came from description, keep description as-is
    let description = raw
        .get("description")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    Some(Skill {
        id: skill_id,
        name,
        description,
        icon: raw
            .get("icon")
            .or(raw.get("emoji"))
            .and_then(|v| v.as_str())
            .unwrap_or("\u{1F4E6}")
            .to_string(),
        system_prompt,
        enabled_tools,
        temperature: raw
            .get("temperature")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.7),
        max_tokens: raw
            .get("max_tokens")
            .and_then(|v| v.as_u64())
            .unwrap_or(4096) as u32,
        color: raw
            .get("color")
            .and_then(|v| v.as_str())
            .unwrap_or("#6366f1")
            .to_string(),
        category: raw
            .get("category")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        format_tag: raw
            .get("format")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
    })
}

fn skill_from_create(id: String, body: SkillCreate) -> Skill {
    Skill {
        id,
        name: body.name,
        description: body.description,
        icon: body.icon,
        system_prompt: body.system_prompt,
        enabled_tools: body.enabled_tools,
        temperature: body.temperature,
        max_tokens: body.max_tokens,
        color: body.color,
        category: body.category,
        format_tag: body.format_tag,
    }
}

/// Convert a Skill to a YAML-serializable Value, excluding the `id` field
/// (the id is encoded in the filename).
fn skill_to_yaml_value(skill: &Skill) -> Value {
    let mut map = serde_json::Map::new();
    map.insert("name".into(), json!(skill.name));
    map.insert("description".into(), json!(skill.description));
    map.insert("icon".into(), json!(skill.icon));
    map.insert("system_prompt".into(), json!(skill.system_prompt));
    map.insert("enabled_tools".into(), json!(skill.enabled_tools));
    map.insert("temperature".into(), json!(skill.temperature));
    map.insert("max_tokens".into(), json!(skill.max_tokens));
    map.insert("color".into(), json!(skill.color));
    if !skill.category.is_empty() {
        map.insert("category".into(), json!(skill.category));
    }
    if !skill.format_tag.is_empty() {
        map.insert("format_tag".into(), json!(skill.format_tag));
    }
    Value::Object(map)
}
